File: cpp/heat2d.cpp
/*
 * 2D Heat equation: u_t = alpha*(u_xx + u_yy)
 * Domain: [0,1]^2, Dirichlet BC=0, IC: sin(pi*x)*sin(pi*y)
 * Exact: exp(-2*alpha*pi^2*t)*sin(pi*x)*sin(pi*y)
 * Method: explicit FTCS
 * Output: heat2d_result.csv (x, y, u_numerical, u_exact, error)
 */

#include <cmath>
#include <cstdio>
#include <vector>
#include <fstream>
#include <string>

static const double PI    = std::acos(-1.0);
static const double ALPHA = 1.0;
static const int    N     = 50;
static const double T_END = 0.1;
static const double R     = 0.2;    // alpha*dt/dx^2

int idx(int i, int j) { return i * (N+2) + j; }

int main()
{
    int    M  = N + 2;
    double dx = 1.0 / (N + 1);
    double dt = R * dx * dx / ALPHA;
    int    steps = static_cast<int>(T_END / dt) + 1;

    std::vector<double> x(M);
    for (int i = 0; i < M; ++i) x[i] = i * dx;

    int total = M * M;
    std::vector<double> u(total, 0.0), u_new(total, 0.0);

    // initial condition
    for (int i = 0; i < M; ++i)
        for (int j = 0; j < M; ++j)
            u[idx(i,j)] = std::sin(PI*x[i]) * std::sin(PI*x[j]);
    // zero boundary (already 0 at i=0,M-1,j=0,M-1)

    // time march
    for (int n = 0; n < steps; ++n) {
        for (int i = 1; i <= N; ++i)
            for (int j = 1; j <= N; ++j) {
                double lap = (u[idx(i+1,j)] + u[idx(i-1,j)]
                            + u[idx(i,j+1)] + u[idx(i,j-1)]
                            - 4.0*u[idx(i,j)]) / (dx*dx);
                u_new[idx(i,j)] = u[idx(i,j)] + ALPHA*dt*lap;
            }
        // copy interior, keep boundary=0
        for (int i = 1; i <= N; ++i)
            for (int j = 1; j <= N; ++j)
                u[idx(i,j)] = u_new[idx(i,j)];
    }

    // exact & error
    double t_final = (steps-1)*dt;
    double decay   = std::exp(-2.0*ALPHA*PI*PI*t_final);
    double l2 = 0.0;
    for (int i = 1; i <= N; ++i)
        for (int j = 1; j <= N; ++j) {
            double ex = decay * std::sin(PI*x[i]) * std::sin(PI*x[j]);
            double e  = u[idx(i,j)] - ex;
            l2 += e*e * dx*dx;
        }
    l2 = std::sqrt(l2);
    std::printf("2D Heat  N=%d  T=%.2f  L2 error = %.3e\n", N, T_END, l2);

    std::ofstream f("heat2d_result.csv");
    f << "x,y,u_numerical,u_exact,error\n";
    for (int i = 0; i < M; ++i)
        for (int j = 0; j < M; ++j) {
            double ex = decay * std::sin(PI*x[i]) * std::sin(PI*x[j]);
            f << x[i] << "," << x[j] << ","
              << u[idx(i,j)] << "," << ex << ","
              << (u[idx(i,j)]-ex) << "\n";
        }
    std::printf("Saved heat2d_result.csv\n");
    return 0;
}

File: fortran/heat2d.f90
! 2D Heat equation: u_t = alpha*(u_xx+u_yy)
! Domain: [0,1]^2, Dirichlet BC=0, IC: sin(pi*x)*sin(pi*y)
! Exact: exp(-2*alpha*pi^2*t)*sin(pi*x)*sin(pi*y)
! Method: explicit FTCS
! Output: heat2d_result.csv

program heat2d
    implicit none

    real(8), parameter :: alpha = 1.0d0
    integer, parameter :: N     = 50
    real(8), parameter :: T_END = 0.1d0
    real(8), parameter :: R     = 0.2d0

    real(8), parameter :: PI = acos(-1.0d0)

    real(8) :: dx, dt, t_final, l2, e, decay, ex
    integer :: M, steps, n, i, j
    real(8), allocatable :: x(:), u(:,:), u_new(:,:)

    M  = N + 2
    dx = 1.0d0 / (N + 1)
    dt = R * dx * dx / alpha
    steps = int(T_END / dt) + 1

    allocate(x(M), u(M,M), u_new(M,M))

    do i = 1, M
        x(i) = (i-1) * dx
    end do

    ! initial condition
    do j = 1, M
        do i = 1, M
            u(i,j) = sin(PI*x(i)) * sin(PI*x(j))
        end do
    end do
    u(1,:) = 0.0d0; u(M,:) = 0.0d0
    u(:,1) = 0.0d0; u(:,M) = 0.0d0

    ! time march
    do n = 1, steps
        do j = 2, M-1
            do i = 2, M-1
                u_new(i,j) = u(i,j) + R*(u(i+1,j)+u(i-1,j) &
                                        +u(i,j+1)+u(i,j-1)-4.0d0*u(i,j))
            end do
        end do
        u_new(1,:) = 0.0d0; u_new(M,:) = 0.0d0
        u_new(:,1) = 0.0d0; u_new(:,M) = 0.0d0
        u = u_new
    end do

    ! exact & L2
    t_final = (steps - 1) * dt
    decay   = exp(-2.0d0 * alpha * PI**2 * t_final)
    l2 = 0.0d0
    do j = 2, M-1
        do i = 2, M-1
            ex = decay * sin(PI*x(i)) * sin(PI*x(j))
            e  = u(i,j) - ex
            l2 = l2 + e*e * dx*dx
        end do
    end do
    l2 = sqrt(l2)
    write(*,'(A,I4,A,F6.3,A,ES10.3)') &
        '2D Heat  N=', N, '  T=', T_END, '  L2 error = ', l2

    ! save CSV
    open(unit=10, file='heat2d_result.csv', status='replace')
    write(10,'(A)') 'x,y,u_numerical,u_exact,error'
    do j = 1, M
        do i = 1, M
            ex = decay * sin(PI*x(i)) * sin(PI*x(j))
            write(10,'(5(ES14.6,","))') x(i), x(j), u(i,j), ex, u(i,j)-ex
        end do
    end do
    close(10)
    write(*,*) 'Saved heat2d_result.csv'

    deallocate(x, u, u_new)
end program heat2d
